registry/client/repository.go

```diff
@@ -632,6 +635,50 @@ func (ms *manifests) Delete(ctx context.Context, dgst digest.Digest) error {
 	return HandleErrorResponse(resp)
 }
 
+func (ms *manifests) digestFromResponse(ctx context.Context, resp *http.Response, manifestURL string) (string, error) {
+	// The specs are ambivalent on Docker-Content-Digest header: on the one hand, the docs at https://docs.docker.com/registry/spec/api/#digest-header
+	// mention that "any response may include a Docker-Content-Digest header"  and on the other hand, the docs at
+	// https://docs.docker.com/registry/spec/api/#put-manifest say that a PUT request should return a Docker-Content-Digest header.
+	//
+	// Because a registry may be accessed over proxies and/or load-balancers (often happens in large corporate environments), there's no guarantee
+	// that the proxies won't prune the Docker-Content-Digest away which would result in a digest mismatch error.
+	//
+	// This function first tries to get the digest from the response headers - if that fails it falls back on getting
+	// the digest from the response body. As a last fallback, a HEAD request is made.
+	// The function may return an empty digest - it is up to the caller to determine if not receiving the digest is an error.
+	dgstHeader := resp.Header.Get("Docker-Content-Digest")
+	if dgstHeader != "" {
+		return dgstHeader, nil
+	}
+
+	// check for digest in response body if no Docker-Content-Digest header is set
+	mt := resp.Header.Get("Content-Type")
+	if mt != "" {
+		body, err := io.ReadAll(resp.Body)
+		if err == nil {
+			_, desc, err := distribution.UnmarshalManifest(mt, body)
+			if err != nil {
+				return "", err
+			}
+			dgstHeader = desc.Digest.String()
+		}
+	}
```

Comment:
Is this correct? Does the `PUT` response contain the uploaded data that was sent by the client? I think the response body would be either empty (only a static header) or an error-response in JSON format.

Reply:
Sonatype Nexus returns the digest as part of the body. I think some other registries do as well - see #3792 for context.